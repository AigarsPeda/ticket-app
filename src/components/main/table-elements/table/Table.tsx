import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import moment from "moment";
import "./Table.scss";
import { RootState } from "../../../../redux/reducers";
import { ITicket } from "../../../../interfaces/interfaces";

const TABLE_HEAD = [
  "ID",
  "Full Name",
  "Subject",
  "Priority",
  "Status",
  "Created",
  "Completed",
  "Action"
];

type Props = ReturnType<typeof mapStateToProps> & typeof mapDispatchToProps;

const Table: React.FC<Props> = (props) => {
  const { tickets } = props;
  const [tableTickets, setTableTickets] = useState<ITicket[]>(tickets);

  useEffect(() => {
    const tableEntries = tickets.slice(0, 3);
    setTableTickets(tableEntries);
  }, [tickets]);

  return (
    <div className="col-sm-12 table-responsive">
      <table className="table table-centered mb-0" id="ticketTable">
        <thead className="font-14 text-light">
          <tr>
            {TABLE_HEAD.map((tableHead, index) => {
              return (
                <th key={index} className="font-weight-medium">
                  {tableHead}&nbsp;
                  <i className="fas fa-angle-up icon"></i>
                </th>
              );
            })}
          </tr>
        </thead>
        <tbody className="font-14 text-light">
          {tableTickets.map((ticket) => {
            return (
              <tr key={ticket._id}>
                <td>{ticket.tickedId}</td>
                <td>{ticket.fullName}</td>
                <td>{ticket.subject}</td>
                <td>
                  {ticket.priority === "High" ? (
                    <span className="badge badge-danger">
                      {ticket.priority}
                    </span>
                  ) : ticket.priority === "Medium" ? (
                    <span className="badge badge-warning">
                      {ticket.priority}
                    </span>
                  ) : (
                    <span className="badge badge-secondary">
                      {ticket.priority}
                    </span>
                  )}
                </td>
                <td>
                  {ticket.status === "Open" ? (
                    <span className="badge badge-danger">{ticket.status}</span>
                  ) : (
                    <span className="badge badge-success">{ticket.status}</span>
                  )}
                </td>
                <td>{moment(ticket.created).format("DD/MM/YYYY")}</td>
                <td>{moment(ticket.dueDate).format("DD/MM/YYYY")}</td>
                <td>
                  <>
                    <a className="btn text-white btn-sm">
                      <i className="fas fa-trash"></i>
                    </a>
                    <a className="btn text-white btn-sm">
                      <i className="fas fa-check"></i>
                    </a>
                    <a className="btn text-white btn-sm">
                      <i className="fas fa-pencil-alt"></i>
                    </a>
                  </>
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
};

const mapStateToProps = (state: RootState) => ({
  tickets: state.tickets.tickets
});

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(Table);
