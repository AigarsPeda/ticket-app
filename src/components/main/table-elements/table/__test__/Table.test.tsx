import React from "react";

import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";

import renderer from "react-test-renderer";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Table from "../Table";
import Enzyme, { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter() });

const ticketDataArray = [
  {
    fullName: "Aigars",
    email: "aigarspeda@gmail.com",
    subject: "Test_1",
    description: "test",
    department: "IT",
    priority: "Low",
    created: "2020-08-09T17:05:29.739Z",
    dueDate: "2020-08-09T17:05:29.739Z",
    status: "Open",
    tickedId: "12334354454",
    _id: "121132334545465"
  },
  {
    fullName: "Oskars",
    email: "oskars@gmail.com",
    subject: "Test_2",
    description: "test_2",
    department: "IT",
    priority: "Low",
    created: "2020-08-09T17:05:29.739Z",
    dueDate: "2020-08-09T17:05:29.739Z",
    status: "Close",
    tickedId: "12334354455",
    _id: "121132334545467"
  }
];

let state: any;
let mockStore: any;
let store: any;
let history: any;

beforeEach(() => {
  state = {
    auth: {
      isAuthenticated: true
    },
    tickets: {
      tickets: ticketDataArray
    },
    entries: {
      entries: 5
    },
    user: {
      user: {
        // date: "2020-08-13T06:45:10.963Z",
        // role: "User",
        // tickets: [],
        // username: "Aigarsp1",
        // __v: 0,
        // _id: "5f34e176cd9f5238ec3fe1ba"
      }
    }
  };
  mockStore = configureMockStore();
  store = mockStore(() => state);
  history = createMemoryHistory();
});

describe("<Table />", () => {
  it("rendering a component and expect text in it", () => {
    // const defaultProps: any = {
    //   history: {
    //     replace: jest.fn()
    //   }
    // };
    const { getByTestId } = render(
      <Provider store={store}>
        <Router history={history}>
          <Table
          // {...defaultProps}
          />
        </Router>
      </Provider>
    );
    expect(getByTestId("tableComponent")).toHaveTextContent("Full Name");
    expect(getByTestId("tableComponent")).toHaveTextContent("Subject");
  });

  it("testing table", () => {
    const container = mount(
      <Provider store={store}>
        <Table />
      </Provider>
    );
    // console.log(container.debug());

    const tbody = container.find("tbody");
    const tr = tbody.find("tr");

    expect(container).toHaveLength(1);
    expect(container.find("table")).toHaveLength(1);
    expect(container.find("tbody")).toHaveLength(1);
    expect(container.find("th")).toHaveLength(8);
    expect(tr).toHaveLength(ticketDataArray.length);
  });

  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router history={history}>
            <Table />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
