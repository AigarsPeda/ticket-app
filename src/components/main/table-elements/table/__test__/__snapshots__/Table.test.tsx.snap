// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> matches snapshot 1`] = `
<div
  className="col-sm-12 table-responsive"
  data-testid="tableComponent"
>
  <table
    className="table table-centered mb-0"
    id="ticketTable"
  >
    <thead
      className="font-14 text-light"
    >
      <tr>
        <th
          className="font-weight-medium"
        >
          ID
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Full Name
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Subject
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Priority
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Status
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Created
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Completed
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
        <th
          className="font-weight-medium"
        >
          Action
           
          <i
            className="fas fa-angle-up icon"
          />
        </th>
      </tr>
    </thead>
    <tbody
      className="font-14 text-light"
    >
      <tr>
        <td>
          12334354454
        </td>
        <td>
          Aigars
        </td>
        <td>
          Test_1
        </td>
        <td>
          <span
            className="badge badge-secondary"
          >
            Low
          </span>
        </td>
        <td>
          <span
            className="badge badge-danger"
          >
            Open
          </span>
        </td>
        <td>
          09/08/2020
        </td>
        <td>
          09/08/2020
        </td>
        <td>
          <a
            className="btn text-white btn-sm"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash"
            />
          </a>
          <a
            className="btn text-white btn-sm"
            onClick={[Function]}
          >
            <i
              className="fas fa-check"
            />
          </a>
          <a
            className="btn text-white btn-sm"
            onClick={[Function]}
          >
            <i
              className="fas fa-pencil-alt"
            />
          </a>
        </td>
      </tr>
      <tr>
        <td>
          12334354455
        </td>
        <td>
          Oskars
        </td>
        <td>
          Test_2
        </td>
        <td>
          <span
            className="badge badge-secondary"
          >
            Low
          </span>
        </td>
        <td>
          <span
            className="badge badge-success"
          >
            Close
          </span>
        </td>
        <td>
          09/08/2020
        </td>
        <td>
          09/08/2020
        </td>
        <td>
          <a
            className="btn text-white btn-sm"
            onClick={[Function]}
          >
            <i
              className="fas fa-trash"
            />
          </a>
          <a
            className="btn text-white btn-sm disabled"
            onClick={[Function]}
          >
            <i
              className="fas fa-check"
            />
          </a>
          <a
            className="btn text-white btn-sm disabled"
            onClick={[Function]}
          >
            <i
              className="fas fa-pencil-alt"
            />
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
